### ISSUE TITLE
Proposal: multi-task models

### ISSUE THREAD

**Issue by colobas** [AllenCellModeling/cyto-dl#1]
# Intro
After talking to @benjijamorris we realized we were both interested in the ability of training multi-task models. @benjijamorris has been developing this [in a separate repo](https://github.com/aics-int/transfer_function_pytorch_lightning) and demonstrated interest in porting some of this functionality into serotiny or a serotiny-based project. I had started toying with a similar idea [in another repo](https://github.com/AllenCellModeling/universal-label-free)

We decided to create this repository to accomodate further development in this direction, and to host other experimental solutions for im2im problems. Code that becomes general enough can graduate to `serotiny`.

### Single task per batch vs multi-task per batch
The main difference between my initial approach and @benjijamorris 's is that he considered a case where in a single training step there are multiple tasks for the same input, whereas I considered a case where each batch contains a single task per input. Likely, the former can subsume the latter.

# Moving forward

We can use the general structure I started up with, and improve it to encompass @benjijamorris 's additional functionality.

### Existing boilerplate

- [General model class for multi task im2im](https://github.com/AllenCellModeling/aics-im2im/blob/main/aics_im2im/models/multi_task.py)
- [Example unfinished & untested config for this class](https://github.com/AllenCellModeling/aics-im2im/blob/main/aics_im2im/config/model/multi_task.yaml)
- Some of @benjijamorris 's defined networks and utils were adapted and pulled into this repo [here](https://github.com/AllenCellModeling/aics-im2im/tree/main/aics_im2im/networks) and [here](https://github.com/AllenCellModeling/aics-im2im/tree/main/aics_im2im/networks)

## Next steps

- Implement the training loop [here](https://github.com/AllenCellModeling/aics-im2im/blob/ff9d891e4ecfccfcd5b37fc97b85e010e2ada6b6/aics_im2im/models/multi_task.py#L64) (might not be much more than copying the existing loop)
- Implement/adapt/port the datamodule
- Move some the validation / metric calculation to separate callbacks
- Add a `predict_step` for inference
